core/src/main/java/io/grpc/LoadBalancer2.java

```diff
@@ -0,0 +1,327 @@
+   *
+   * @param error a non-OK status
+   */
+  public abstract void handleNameResolutionError(Status error);
+
+  /**
+   * Handles a state change on a Subchannel.
+   *
+   * @param subchannel the involved Subchannel
+   * @param stateInfo the new state
+   */
+  public abstract void handleSubchannelState(
+      Subchannel subchannel, ConnectivityStateInfo stateInfo);
+
+  /**
+   * The channel asks the load-balancer to shutdown.  No more callbacks will be called after this
+   * method.  The implementation should shutdown all Subchannels and OOB channels, and do any other
+   * cleanup as necessary.
+   */
+  public abstract void shutdown();
+
+  /**
+   * The main balancing logic.  It <strong>must be thread-safe</strong>. Typically it should only
+   * synchronize on its own state, and void synchronzing with the LoadBalancer's state.
+   */
+  @ThreadSafe
+  public abstract static class SubchannelPicker {
+    /**
+     * Make a balancing decision for a new RPC.
+     *
+     * @param affinity the affinity attributes provided via {@link CallOptions#withAffinity}
+     * @param headers the headers container of the RPC. It can be mutated within this method.
+     */
+    public abstract PickResult pickSubchannel(Attributes affinity, Metadata headers);
+  }
+
+  /**
+   * A balancing decision made by {@link SubchannelPicker} for an RPC.
+   */
+  @Immutable
+  public static final class PickResult {
+    // A READY channel, or null
+    @Nullable private final Subchannel subchannel;
+    // An error to be propagated to the application if subchannel == null
+    // Or OK if there is no error.
+    // subchannel being null and error being OK means RPC needs to wait
+    private final Status status;
+
+    private PickResult(Subchannel subchannel, Status status) {
+      this.subchannel = subchannel;
+      this.status = Preconditions.checkNotNull(status, "status");
+    }
+
+    /**
+     * A decision to proceed the RPC on a Subchannel.  The state of the Subchannel is supposed to be
+     * {@link ConnectivityState#READY}.  However, a non-READY Subchannel will not fail the RPC,
+     * but will only leave it buffered.
+     */
+    public static PickResult withSubchannel(Subchannel subchannel) {
+      return new PickResult(Preconditions.checkNotNull(subchannel, "subchannel"), Status.OK);
+    }
+
+    /**
+     * A decision to report a connectivity error to the RPC.  If the RPC is {@link
+     * CallOptions#withWaitForReady wait-for-ready}, it will stay buffered.  Otherwise, it will fail
+     * with the given error.
+     *
+     * @param error the error status.  Must not be OK.
+     */
+    public static PickResult withError(Status error) {
+      Preconditions.checkArgument(!error.isOk(), "error status shouldn't be OK");
+      return new PickResult(null, error);
+    }
+
+    /**
+     * No decision could be made.  The RPC will stay buffered.
+     */
+    public static PickResult withNothingYet() {
```

Comment:
That makes sense. Maybe `withNoResult()` would work better? I like `pending()` the most, but it's missing `with` and I'd like to to be consistent with other methods. Heh.